pkg/apis/extensions/v1beta1/types.go

```diff
@@ -286,6 +290,47 @@ type DeploymentStatus struct {
 
 	// Total number of non-terminated pods targeted by this deployment that have the desired template spec.
 	UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+	// Conditions represent the latest available observations of a deployment's current state.
+	Conditions []DeploymentCondition `json:"conditions,omitempty"`
+}
+
+type DeploymentConditionType string
+
+// These are valid conditions of a deployment.
+const (
+	// DeploymentProgressing means the deployment is progressing.
+	DeploymentProgressing DeploymentConditionType = "Progressing"
+	// DeploymentComplete means the deployment has completed progressing.
+	DeploymentComplete DeploymentConditionType = "Complete"
+	// DeploymentFailed means the deployment has failed progressing.
+	DeploymentFailed DeploymentConditionType = "Failed"
+)
+
+// DeploymentCondition describes the state of a deployment at a certain point.
+type DeploymentCondition struct {
+	// Type of deployment condition.
+	Type DeploymentConditionType `json:"type"`
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+	// LastProbeTime is the last time the condition was checked.
+	LastProbeTime unversioned.Time `json:"lastProbeTime,omitempty"`
```

Comment:
Looking at the semantics of these timestamps closely, now I believe we need `LastProbeTime` instead of `LastTransitionTime` for Deployments. Currently we have three different types:

* Available

Deployment readiness/availability is different from pod readiness/availability. The latter is fixed once we pin down to a True state whereas the former is looking over other resources so it needs to be updated every time it probes.

* Progressing

Every time we probe the deployment for progress we need to store the time we probed so next time we know how much time has passed since last time progress was reported. That's the way we use the condition actually atm.

* ReplicaFailure

This condition is added/removed accordingly to the failures found in deployment pods. `LastTransitionTime` doesn't make sense because we remove the Condition altogether once a Deployment has no pod failures. This doesn't happen elsewhere in Condiitons for other resources that I know of. My thinking is that since there is no failure, we shouldn't overload the deployment status. `LastProbeTime` doesn't make sense either since I want to keep the first time we encountered a replica failure but at the end of the day I think the timestamp here is not so important and hopefully the name won't be confusing for users.

All that being said, I am going to switch `LastTransitionTime` to `LastProbeTime` for Deployment Conditions. Let me know if you have any arguments against the change.